File: python/timestep/ag_ui_server.py
import os
import json
from typing import AsyncGenerator
from ag_ui.core import (
    RunAgentInput,
    BaseEvent,
    EventType,
    RunStartedEvent,
    RunFinishedEvent,
    RunErrorEvent,
    TextMessageStartEvent,
    TextMessageChunkEvent,
    TextMessageEndEvent,
    ToolCallStartEvent,
    ToolCallArgsEvent,
    ToolCallResultEvent,
    ToolCallEndEvent,
)
from agents import Agent, Runner, function_tool, RunConfig
from multi_provider import MultiProvider, MultiProviderMap
from ollama_model_provider import OllamaModelProvider
from agents import OpenAIProvider
from mcp_server_proxy import fetch_mcp_tools


class TimestepAgent:
    """AG-UI compatible agent that wraps the OpenAI agents framework."""

    def __init__(self, model_id: str = None, openai_use_responses: bool = False):
        self.model_id = model_id or os.getenv("MODEL_ID") or "ollama/gpt-oss:120b-cloud"
        self.openai_use_responses = openai_use_responses

    async def run(self, input: RunAgentInput) -> AsyncGenerator[BaseEvent, None]:
        """
        Run the agent and yield AG-UI events.

        Args:
            input: RunAgentInput containing threadId, runId, and messages

        Yields:
            BaseEvent objects conforming to AG-UI protocol
        """
        try:
            # Emit RUN_STARTED event
            yield RunStartedEvent(thread_id=input.thread_id, run_id=input.run_id)

            # Run the agent and emit events
            async for event in self._run_agent(input):
                yield event

            # Emit RUN_FINISHED event
            yield RunFinishedEvent(thread_id=input.thread_id, run_id=input.run_id)

        except Exception as error:
            # Emit RUN_ERROR event
            yield RunErrorEvent(message=str(error))
            raise

    async def _run_agent(self, input: RunAgentInput) -> AsyncGenerator[BaseEvent, None]:
        """Internal method to run the agent and map events to AG-UI protocol."""

        # Set up model providers
        model_provider_map = MultiProviderMap()

        model_provider_map.add_provider(
            "ollama",
            OllamaModelProvider(api_key=os.getenv("OLLAMA_API_KEY")),
        )

        model_provider_map.add_provider(
            "anthropic",
            OpenAIProvider(
                api_key=os.getenv("ANTHROPIC_API_KEY"),
                base_url="https://api.anthropic.com/v1/",
                use_responses=False,
            ),
        )

        model_provider = MultiProvider(
            provider_map=model_provider_map,
            openai_use_responses=self.openai_use_responses,
        )

        run_config = RunConfig(
            model_provider=model_provider,
            trace_include_sensitive_data=True,
            tracing_disabled=False,
        )

        # Configure approval policies (get_weather doesn't require approval for AG-UI)
        require_approval = {
            "never": {"toolNames": ["search_codex_code", "fetch_codex_documentation", "get_weather"]},
            "always": {"toolNames": ["fetch_generic_url_content"]},
        }

        # Fetch built-in tools for weather agent
        print('[MCP] Loading built-in tools...')
        weather_tools = await fetch_mcp_tools(None, True, require_approval)
        print(f'[MCP] Loaded {len(weather_tools)} built-in tools')

        # Create weather agent with built-in tools
        weather_agent = Agent(
            model=self.model_id,
            name="Weather agent",
            instructions="You provide weather information.",
            handoff_description="Handles weather-related queries",
            tools=weather_tools,
        )

        # Fetch remote MCP tools from the codex server
        print('[MCP] Fetching tools from https://gitmcp.io/timestep-ai/timestep...')
        mcp_tools = await fetch_mcp_tools('https://gitmcp.io/timestep-ai/timestep', False, require_approval)
        print(f'[MCP] Loaded {len(mcp_tools)} remote tools')

        # Create main agent with remote MCP tools and weather handoff
        agent = Agent(
            model=self.model_id,
            name="Main Assistant",
            instructions="You are a helpful assistant. For questions about the timestep-ai/timestep repository, use the MCP tools. For weather questions, hand off to the weather agent.",
            tools=mcp_tools,
            handoffs=[weather_agent],
        )

        runner = Runner()

        # Get the last user message as input
        messages = input.messages or []
        user_messages = [m for m in messages if m.role == "user"]
        last_user_message = user_messages[-1] if user_messages else None
        user_input = last_user_message.content if last_user_message else ""

        if not user_input:
            # No user input, just return without doing anything
            return

        result = Runner.run_streamed(agent, user_input, run_config=run_config)

        current_message_id = None
        message_started = False

        async for event in result.stream_events():
            # Check for message output events that contain text
            if event.type == "run_item_stream_event":
                item = event.item

                if item.type == "message_output_item":
                    # Get text content from message
                    from agents import ItemHelpers

                    text_content = ItemHelpers.text_message_output(item)
                    if text_content:
                        # Initialize message ID and send TEXT_MESSAGE_START
                        if not current_message_id:
                            current_message_id = getattr(item, "id", str(id(item)))
                            yield TextMessageStartEvent(message_id=current_message_id)
                            message_started = True

                        # Send TEXT_MESSAGE_CHUNK
                        yield TextMessageChunkEvent(
                            message_id=current_message_id, delta=text_content
                        )

                # Handle tool call events
                elif item.type == "tool_call_item":
                    # Get tool call information from raw_item
                    raw_item = item.raw_item
                    tool_call_id = raw_item.call_id
                    tool_call_name = raw_item.name
                    tool_args = raw_item.arguments

                    yield ToolCallStartEvent(
                        tool_call_id=tool_call_id, tool_call_name=tool_call_name
                    )

                    # Convert args to JSON string (handle both dict and string)
                    if isinstance(tool_args, dict):
                        args_str = json.dumps(tool_args)
                    elif isinstance(tool_args, str):
                        args_str = tool_args
                    else:
                        args_str = str(tool_args)

                    yield ToolCallArgsEvent(tool_call_id=tool_call_id, delta=args_str)

                elif item.type == "tool_call_output_item":
                    # Get tool output information
                    raw_item = item.raw_item
                    tool_call_id = (
                        raw_item["call_id"]
                        if isinstance(raw_item, dict)
                        else raw_item.call_id
                    )
                    result_text = item.output
                    message_id = str(id(item))

                    yield ToolCallResultEvent(
                        message_id=message_id,
                        tool_call_id=tool_call_id,
                        content=result_text,
                    )

                    yield ToolCallEndEvent(tool_call_id=tool_call_id)

        # Send TEXT_MESSAGE_END if we started a message
        if message_started and current_message_id:
            yield TextMessageEndEvent(message_id=current_message_id)

File: typescript/timestep/ag_ui_server.ts
import {
  AbstractAgent,
  RunAgentInput,
  EventType,
  BaseEvent,
} from "@ag-ui/client"
import { Observable } from "rxjs"
import { z } from 'zod'
import { Agent, Runner, tool, OpenAIProvider, RunToolApprovalItem } from '@openai/agents'
import { RunConfig, RunState } from '@openai/agents-core'
import { MultiProvider, MultiProviderMap } from './multi_provider'
import { OllamaModelProvider } from './ollama_model_provider'
import { fetchMcpTools } from './mcp_server_proxy'
import { writeFileSync, readFileSync, existsSync, mkdirSync } from 'node:fs'

export class TimestepAgent extends AbstractAgent {
  private modelId: string
  private openaiUseResponses: boolean

  constructor(modelId?: string, openaiUseResponses: boolean = false) {
    super()
    this.modelId = modelId ?? process.env.MODEL_ID ?? 'ollama/gpt-oss:120b-cloud'
    this.openaiUseResponses = openaiUseResponses
  }

  protected run(input: RunAgentInput): Observable<BaseEvent> {
    return new Observable<BaseEvent>((observer) => {
      observer.next({
        type: EventType.RUN_STARTED,
        threadId: input.threadId,
        runId: input.runId,
      } as any)

      this.runAgent(input, observer)
        .then(() => {
          observer.next({
            type: EventType.RUN_FINISHED,
            threadId: input.threadId,
            runId: input.runId,
          } as any)
          observer.complete()
        })
        .catch((error) => {
          observer.next({
            type: EventType.RUN_ERROR,
            message: error.message,
          } as any)
          observer.error(error)
        })
    })
  }

  private async runAgent(input: RunAgentInput, observer: any) {
    // Set up model providers
    const modelProviderMap = new MultiProviderMap()

    modelProviderMap.addProvider("ollama", new OllamaModelProvider({
      apiKey: process.env.OLLAMA_API_KEY,
    }))

    modelProviderMap.addProvider("anthropic", new OpenAIProvider({
      apiKey: process.env.ANTHROPIC_API_KEY,
      baseURL: "https://api.anthropic.com/v1/",
      useResponses: false
    }))

    const modelProvider = new MultiProvider({
      provider_map: modelProviderMap,
      openai_use_responses: this.openaiUseResponses,
    })

    const runConfig: RunConfig = {
      modelProvider: modelProvider,
      traceIncludeSensitiveData: true,
      tracingDisabled: false,
    }

    // Configure approval policies (get_weather now requires approval for AG-UI testing)
    const requireApproval = {
      never: { toolNames: ['search_codex_code', 'fetch_codex_documentation'] },
      always: { toolNames: ['get_weather', 'fetch_generic_url_content'] },
    }

    // Fetch built-in tools for weather agent
    console.log('[MCP] Loading built-in tools...')
    const weatherTools = await fetchMcpTools(null, true, requireApproval)
    console.log(`[MCP] Loaded ${weatherTools.length} built-in tools`)

    // Create weather agent with built-in tools
    const weatherAgent = new Agent({
      model: this.modelId,
      name: 'Weather agent',
      instructions: 'You provide weather information.',
      handoffDescription: 'Handles weather-related queries',
      tools: weatherTools,
    })

    // Fetch remote MCP tools from the codex server
    console.log('[MCP] Fetching tools from https://gitmcp.io/timestep-ai/timestep...')
    const mcpTools = await fetchMcpTools('https://gitmcp.io/timestep-ai/timestep', false, requireApproval)
    console.log(`[MCP] Loaded ${mcpTools.length} remote tools`)

    // Create main agent with remote MCP tools and weather handoff
    const agent = new Agent({
      model: this.modelId,
      name: 'Main Assistant',
      instructions:
        'You are a helpful assistant. For questions about the timestep-ai/timestep repository, use the MCP tools. For weather questions, hand off to the weather agent.',
      tools: mcpTools,
      handoffs: [weatherAgent],
    })

    const runner = new Runner(runConfig)

    // Ensure data directory exists for state persistence
    mkdirSync('data', { recursive: true })

    // Check if this is a resume from an interrupt
    let agentInput: string | RunState<undefined, Agent<unknown, "text">>
    let isResumingFromInterrupt = false

    if (input.resume?.interruptId) {
      // Load saved state from disk
      const stateFilename = `data/${input.threadId}.state.json`
      if (existsSync(stateFilename)) {
        console.log('[AG-UI] Resuming from interrupt...')
        const raw = readFileSync(stateFilename, 'utf8')
        const savedState = JSON.parse(raw)
        const runState = await RunState.fromString(agent, JSON.stringify(savedState))

        // Find the tool approval item
        const approvalItem = runState._generatedItems
          .filter(item => item.type === 'tool_approval_item')
          .find(item => (item as RunToolApprovalItem).rawItem?.id === input.resume?.interruptId) as RunToolApprovalItem | undefined

        if (approvalItem) {
          // Apply the approval decision from resume payload
          const approved = input.resume.payload?.approved ?? false
          const alwaysApprove = input.resume.payload?.alwaysApprove ?? false
          const alwaysReject = input.resume.payload?.alwaysReject ?? false

          if (approved) {
            runState.approve(approvalItem, { alwaysApprove })
            console.log('[AG-UI] Tool approved')
          } else {
            runState.reject(approvalItem, { alwaysReject })
            console.log('[AG-UI] Tool rejected')
          }

          agentInput = runState as RunState<undefined, Agent<unknown, "text">>
          isResumingFromInterrupt = true
        } else {
          console.error('[AG-UI] Warning: Could not find approval item for interrupt ID')
          // Fall back to normal execution
          const messages = input.messages ?? []
          const userMessages = messages.filter(m => m.role === 'user')
          const lastUserMessage = userMessages[userMessages.length - 1]
          agentInput = lastUserMessage?.content ?? ''
        }
      } else {
        console.error('[AG-UI] Warning: No saved state found for resume')
        // Fall back to normal execution
        const messages = input.messages ?? []
        const userMessages = messages.filter(m => m.role === 'user')
        const lastUserMessage = userMessages[userMessages.length - 1]
        agentInput = lastUserMessage?.content ?? ''
      }
    } else {
      // Normal execution - get the last user message as input
      const messages = input.messages ?? []
      const userMessages = messages.filter(m => m.role === 'user')
      const lastUserMessage = userMessages[userMessages.length - 1]
      agentInput = lastUserMessage?.content ?? ''

      if (!agentInput) {
        // No user input, just return without doing anything
        return
      }
    }

    const stream = await runner.run(
      agent,
      agentInput,
      { stream: true },
    )

    let currentMessageId: string | null = null
    let messageStarted = false

    for await (const chunk of stream) {
      if ('name' in chunk) {
        // Handle tool approval requests - save state and emit interrupt
        if (chunk.name === 'tool_approval_requested') {
          const toolApprovalItem = (chunk as any).item
          if (toolApprovalItem?.rawItem?.providerData?.function?.name) {
            const toolName = toolApprovalItem.rawItem.providerData.function.name
            const toolArguments = toolApprovalItem.rawItem.providerData.function.arguments
            const toolAgent = toolApprovalItem.agent?.name
            const interruptId = toolApprovalItem.rawItem.id || Date.now().toString()

            console.log('[AG-UI] Tool approval required, saving state...')

            // Save state to disk
            const state = (stream as any).state as RunState<any, any>
            const stateFilename = `data/${input.threadId}.state.json`
            try {
              const savedState = JSON.parse(JSON.stringify(state))
              writeFileSync(stateFilename, JSON.stringify(savedState, null, 2))
              console.log(`[AG-UI] State saved to ${stateFilename}`)
            } catch (err) {
              console.error('[AG-UI] Failed to serialize stream state:', err)
              throw err
            }

            // Emit RUN_FINISHED with interrupt outcome
            observer.next({
              type: EventType.RUN_FINISHED,
              threadId: input.threadId,
              runId: input.runId,
              outcome: 'interrupt',
              interrupt: {
                id: interruptId,
                reason: 'tool_approval_required',
                payload: {
                  toolName,
                  toolArguments,
                  agentName: toolAgent,
                }
              }
            } as any)

            // Complete the stream
            return
          }
        }

        // Check for message output events that contain text
        else if (chunk.name === 'message_output_created') {
          const messageItem = (chunk as any).item

          // Check if this is a message with text content
          if (messageItem?.rawItem?.content) {
            const content = messageItem.rawItem.content

            // Initialize message ID and send TEXT_MESSAGE_START
            if (!currentMessageId) {
              currentMessageId = messageItem.rawItem.id || Date.now().toString()
              observer.next({
                type: EventType.TEXT_MESSAGE_START,
                messageId: currentMessageId,
              } as any)
              messageStarted = true
            }

            // Handle array of content blocks
            if (Array.isArray(content)) {
              for (const part of content) {
                if ((part.type === 'text' || part.type === 'output_text') && part.text) {
                  observer.next({
                    type: EventType.TEXT_MESSAGE_CHUNK,
                    messageId: currentMessageId,
                    delta: part.text,
                  } as any)
                }
              }
            }
            // Handle string content
            else if (typeof content === 'string') {
              observer.next({
                type: EventType.TEXT_MESSAGE_CHUNK,
                messageId: currentMessageId,
                delta: content,
              } as any)
            }
          }
        }

        // Handle tool call events
        else if (chunk.name === 'tool_called') {
          const toolCallItem = (chunk as any).item
          if (toolCallItem?.rawItem?.providerData?.function) {
            const toolCallId = toolCallItem.rawItem.id || Date.now().toString()
            const toolName = toolCallItem.rawItem.providerData.function.name
            const toolArgs = toolCallItem.rawItem.providerData.function.arguments

            observer.next({
              type: EventType.TOOL_CALL_START,
              toolCallId,
              toolName,
            } as any)

            observer.next({
              type: EventType.TOOL_CALL_ARGS,
              toolCallId,
              args: toolArgs,
            } as any)
          }
        }

        else if (chunk.name === 'tool_output') {
          const toolOutputItem = (chunk as any).item
          if (toolOutputItem?.rawItem?.name) {
            const toolCallId = toolOutputItem.rawItem.tool_call_id || Date.now().toString()
            const result = toolOutputItem.output

            observer.next({
              type: EventType.TOOL_CALL_RESULT,
              toolCallId,
              result,
            } as any)

            observer.next({
              type: EventType.TOOL_CALL_END,
              toolCallId,
            } as any)
          }
        }
      }
    }

    // Send TEXT_MESSAGE_END if we started a message
    if (messageStarted && currentMessageId) {
      observer.next({
        type: EventType.TEXT_MESSAGE_END,
        messageId: currentMessageId,
      } as any)
    }

    await stream.completed
  }
}
